src/components/withCurrentReportID.tsx

```diff
@@ -0,0 +1,86 @@
+import React, {createContext, forwardRef, useCallback, useState, useMemo, RefAttributes, ComponentType, ForwardedRef} from 'react';
+import PropTypes from 'prop-types';
+import {NavigationState} from '@react-navigation/native';
+
+import getComponentDisplayName from '../libs/getComponentDisplayName';
+import Navigation from '../libs/Navigation/Navigation';
+
+type CurrentReportIDContextValue = {
+    updateCurrentReportID: (state: NavigationState) => void;
+    currentReportID: string;
+};
+type CurrentReportIDContextProviderProps = {
+    /** Actual content wrapped by this component */
+    children: React.ReactNode;
+};
+
+const CurrentReportIDContext = createContext<CurrentReportIDContextValue | null>(null);
+
+// TODO: Remove when depended components are migrated to TypeScript.
+const withCurrentReportIDPropTypes = {
+    /** Function to update the state */
+    updateCurrentReportID: PropTypes.func.isRequired,
+
+    /** The top most report id */
+    currentReportID: PropTypes.string,
+};
+
+const withCurrentReportIDDefaultProps = {
+    currentReportID: '',
+};
+
+function CurrentReportIDContextProvider(props: CurrentReportIDContextProviderProps) {
+    const [currentReportID, setCurrentReportID] = useState('');
+
+    /**
+     * This function is used to update the currentReportID
+     * @param state root navigation state
+     */
+    const updateCurrentReportID = useCallback(
+        (state: NavigationState) => {
+            setCurrentReportID(Navigation.getTopmostReportId(state) ?? '');
+        },
+        [setCurrentReportID],
+    );
+
+    /**
+     * The context this component exposes to child components
+     * @returns currentReportID to share between central pane and LHN
+     */
+    const contextValue = useMemo(
+        (): CurrentReportIDContextValue => ({
+            updateCurrentReportID,
+            currentReportID,
+        }),
+        [updateCurrentReportID, currentReportID],
+    );
+
+    return <CurrentReportIDContext.Provider value={contextValue}>{props.children}</CurrentReportIDContext.Provider>;
+}
+
+CurrentReportIDContextProvider.displayName = 'CurrentReportIDContextProvider';
+
+export default function withCurrentReportID<TProps extends CurrentReportIDContextValue, TRef>(WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>) {
```

Comment:
Add return type for this function, similar how Bartek did [here](https://github.com/Expensify/App/pull/29426/files)